pallets/block-rewards/src/lib.rs

```diff
@@ -0,0 +1,440 @@
+	///
+	/// NOTE: Noop if any call fails.
+	fn do_advance_epoch() -> Weight {
+		let mut groups = 0;
+		let mut weight_changes = 0;
+		let mut num_joining = 0;
+		let mut num_leaving = 0;
+
+		transactional::with_storage_layer(|| -> DispatchResult {
+			NextEpochChanges::<T>::try_mutate(|changes| -> DispatchResult {
+				ActiveEpochData::<T>::try_mutate(|epoch_data| {
+					num_joining = changes.collators.inc.len();
+					num_leaving = changes.collators.out.len();
+					for leaving in changes.collators.out.drain(..) {
+						Self::do_exit_collator(&leaving)?;
+					}
+					for joining in changes.collators.inc.drain(..) {
+						Self::do_init_collator(&joining)?;
+					}
+
+					groups = T::Rewards::distribute_reward_with_weights(
+						epoch_data.reward,
+						epoch_data.weights.iter().map(|(g, w)| (*g, *w)),
+					)
+					.map(|results| {
+						Self::do_mint_remainder(epoch_data.reward, &results)?;
+						Ok(results.len() as u32)
+					})
+					.and_then(|num_groups| num_groups)?;
+
+					for (&group_id, &weight) in &changes.weights {
+						epoch_data.weights.try_insert(group_id, weight).ok();
+						weight_changes += 1;
+					}
+
+					epoch_data.reward = changes.reward.unwrap_or(epoch_data.reward);
+
+					Self::deposit_event(Event::NewEpoch {
+						reward: epoch_data.reward,
+						last_changes: mem::take(changes),
+					});
+
+					Ok(())
+				})
+			})
+		})
+		.ok();
+
+		// TODO: Apply number of incoming and outgoing collators
+		T::WeightInfo::on_initialize(groups, weight_changes)
+	}
+}
+
+impl<T: Config> sp_runtime::BoundToRuntimeAppPublic for Pallet<T> {
+	type Public = T::AuthorityId;
+}
+
+// Should be instantiated after the original SessionHandler such that current and queued collators are up-to-date for the current session.
+impl<T: Config> OneSessionHandler<T::AccountId> for Pallet<T> {
+	type Key = T::AuthorityId;
+
+	fn on_genesis_session<'a, I: 'a>(_validators: I)
+	where
+		I: Iterator<Item = (&'a T::AccountId, T::AuthorityId)>,
+	{
+		// we don't care.
+	}
+
+	fn on_new_session<'a, I: 'a>(changed: bool, validators: I, queued_validators: I)
+	where
+		I: Iterator<Item = (&'a T::AccountId, T::AuthorityId)>,
+	{
+		// MUST be called before updating collator set changes.
+		// Else the timing is off.
+		let mut weight = Self::do_advance_epoch();
+
+		// Prepare collator set changes for next session.
+		if changed {
+			let current = validators
+				.map(|(acc_id, _)| acc_id.clone())
+				.collect::<Vec<_>>();
+			let next = queued_validators
+				.map(|(acc_id, _)| acc_id.clone())
+				.collect::<Vec<_>>();
+
+			// Prepare for next session
+			if current != next {
+				NextEpochChanges::<T>::mutate(|EpochChanges { collators, .. }| {
+					let inc = next
+						.clone()
+						.into_iter()
+						.filter(|n| !current.iter().any(|curr| curr == n))
+						.collect::<Vec<_>>();
+					let out = current
+						.clone()
+						.into_iter()
+						.filter(|curr| !next.iter().any(|n| n == curr))
+						.collect::<Vec<_>>();
```

Comment:
Could this method be called with a list of validators with duplicated accounts, or Substrate ensures it can not happen?

Reply:
If two collators have the same account, I understand that I want to be rewarded by both collators. Filtering here I'm losing half of the reward (?)

Reply:
> Could this method be called with a list of validators with duplicated accounts, or Substrate ensures it can not happen?

In our case, [this is ensured](https://github.com/paritytech/cumulus/blob/master/pallets/collator-selection/src/lib.rs#L376) by the `pallet-collator-selection` which handles on- and offboarding of collator candidates.

In practice, it also wouldn't make sense to have the same account twice here. In the end, the same entity can easily run multiple collators by using pseudonyms/proxies.

Reply:
Thanks for the info! 🙌🏻